File: Linear/queue/java/Queue.java
/*
 * Author: Christopher Forte
 * Date: March 14, 2026
 * Title: Queue
 * Description: Personal take on a FIFO (first-in, first-out) linear data structure backed by a singly linked list for O(1) enqueue and dequeue.
 */

public class Queue<T> {

    private static class Node<T> {
        T data;
        Node<T> next;
        Node(T data) { this.data = data; this.next = null; }
    }

    private Node<T> head;
    private Node<T> tail;
    private int length;

    // Initializes an empty queue.
    // Input:  none
    // Output: void
    public Queue() { head = null; tail = null; length = 0; }

    // Adds a new value to the back of the queue.
    // Input:  value (T)
    // Output: void
    public void enqueue(T value) {
        Node<T> node = new Node<>(value);
        if (tail != null) tail.next = node;
        tail = node;
        if (head == null) head = tail;
        length++;
    }

    // Removes and returns the front element.
    // Input:  none
    // Output: T
    public T dequeue() {
        if (head == null) throw new java.util.NoSuchElementException("Queue is empty");
        T val = head.data;
        head = head.next;
        if (head == null) tail = null;
        length--;
        return val;
    }

    // Returns the front element without removing it.
    // Input:  none
    // Output: T
    public T front() {
        if (head == null) throw new java.util.NoSuchElementException("Queue is empty");
        return head.data;
    }

    // Returns the back element without removing it.
    // Input:  none
    // Output: T
    public T back() {
        if (tail == null) throw new java.util.NoSuchElementException("Queue is empty");
        return tail.data;
    }

    // Removes all elements from the queue.
    // Input:  none
    // Output: void
    public void clear() { head = null; tail = null; length = 0; }

    // Returns the number of elements in the queue.
    // Input:  none
    // Output: int
    public int size() { return length; }

    // Returns true if the queue holds no elements.
    // Input:  none
    // Output: boolean
    public boolean isEmpty() { return length == 0; }

    // Returns a string of elements from front to back.
    // Input:  none
    // Output: String
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("front [ ");
        Node<T> curr = head;
        while (curr != null) {
            sb.append(curr.data);
            if (curr.next != null) sb.append(", ");
            curr = curr.next;
        }
        return sb.append(" ] back").toString();
    }
}

File: Linear/queue/cpp/Queue.cpp
/*
 * Author: Christopher Forte
 * Date: March 14, 2026
 * Title: Queue
 * Description: Personal take on a FIFO (first-in, first-out) linear data structure backed by a singly linked list.
 */

#pragma once
#include <iostream>
#include <stdexcept>

template <typename T>
class Queue {
private:
    struct Node {
        T data;
        Node* next;
        Node(const T& val) : data(val), next(nullptr) {}
    };

    Node* frontNode;
    Node* backNode;
    int length;

public:
    // Initializes an empty queue.
    // Input:  None
    // Output: None
    Queue() : frontNode(nullptr), backNode(nullptr), length(0) {}

    // Releases all heap-allocated nodes.
    // Input:  None
    // Output: None
    ~Queue() { clear(); }

    // Adds a new value to the back of the queue.
    // Input:  const T& value
    // Output: None
    void enqueue(const T& value) {
        Node* node = new Node(value);
        if (backNode) backNode->next = node;
        backNode = node;
        if (!frontNode) frontNode = backNode;
        length++;
    }

    // Removes and returns the front element.
    // Input:  None
    // Output: T
    T dequeue() {
        if (!frontNode) throw std::underflow_error("Queue is empty");
        Node* node = frontNode;
        T val = node->data;
        frontNode = frontNode->next;
        if (!frontNode) backNode = nullptr;
        delete node;
        length--;
        return val;
    }

    // Returns a const reference to the front element without removing it.
    // Input:  None
    // Output: const T&
    const T& front() const {
        if (!frontNode) throw std::underflow_error("Queue is empty");
        return frontNode->data;
    }

    // Returns a const reference to the back element without removing it.
    // Input:  None
    // Output: const T&
    const T& back() const {
        if (!backNode) throw std::underflow_error("Queue is empty");
        return backNode->data;
    }

    // Removes all elements and resets the queue to empty.
    // Input:  None
    // Output: None
    void clear() {
        while (frontNode) {
            Node* temp = frontNode;
            frontNode = frontNode->next;
            delete temp;
        }
        backNode = nullptr;
        length = 0;
    }

    // Returns the number of elements in the queue.
    // Input:  None
    // Output: int
    int size() const { return length; }

    // Returns true if the queue holds no elements.
    // Input:  None
    // Output: bool
    bool isEmpty() const { return length == 0; }

    // Prints all elements from front to back.
    // Input:  None
    // Output: None (stdout)
    void print() const {
        Node* curr = frontNode;
        std::cout << "front [ ";
        while (curr) {
            std::cout << curr->data;
            if (curr->next) std::cout << ", ";
            curr = curr->next;
        }
        std::cout << " ] back\n";
    }
};
